src/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs

```diff
@@ -168,40 +178,57 @@ public static void CreateErrorResponse(HttpResponse response, RequestContext req
 
-                serviceEx = MsalExceptionFactory.GetServiceException(
-                    msalTokenResponse.Error,
-                    msalTokenResponse.ErrorDescription,
-                    response);
+                    if (msalTokenResponse != null && msalTokenResponse.Error != null)
+                    {
+                        if (CoreErrorCodes.InvalidGrantError.Equals(msalTokenResponse.Error, StringComparison.OrdinalIgnoreCase))
+                        {
+                            throw MsalExceptionFactory.GetUiRequiredException(
+                                CoreErrorCodes.InvalidGrantError,
+                                msalTokenResponse.ErrorDescription,
+                                null,
+                                ExceptionDetail.FromHttpResponse(response));
+                        }
 
-                // For device code flow, AuthorizationPending can occur a lot while waiting
-                // for the user to auth via browser and this causes a lot of error noise in the logs.
-                // So suppress this particular case to an Info so we still see the data but don't 
-                // log it as an error since it's expected behavior while waiting for the user.
-                if (string.Compare(msalTokenResponse.Error, OAuth2Error.AuthorizationPending,
-                        StringComparison.OrdinalIgnoreCase) == 0)
-                {
-                    shouldLogAsError = false;
-                }
+                        serviceEx = MsalExceptionFactory.GetServiceException(
+                            msalTokenResponse.Error,
+                            msalTokenResponse.ErrorDescription,
+                            response);
 
+                        // For device code flow, AuthorizationPending can occur a lot while waiting
+                        // for the user to auth via browser and this causes a lot of error noise in the logs.
+                        // So suppress this particular case to an Info so we still see the data but don't 
+                        // log it as an error since it's expected behavior while waiting for the user.
+                        if (string.Compare(msalTokenResponse.Error, OAuth2Error.AuthorizationPending, StringComparison.OrdinalIgnoreCase) == 0)
+                        {
+                            shouldLogAsError = false;
+                        }
+                    }
+                }
+            }
+            catch (SerializationException) // in the rare case we get an error response we cannot deserialize
+            {
+                serviceEx = MsalExceptionFactory.GetServiceException(
+                    CoreErrorCodes.NonParsableOAuthError,
+                    CoreErrorMessages.NonParsableOAuthError,
+                    response);
             }
-            catch (SerializationException ex)
+            catch (Exception ex)
```

Comment:
The above MsalUiRequiredException is getting caught here, which is why all the tests are having issues for UiRequired stuff showing invalid_grant instead. Maybe that's the same w/dev3x? Currently we're just catching SerializationExceptions, but not all exceptions.

Reply:
will investigate.